core-processor/src/ext.rs

```diff
@@ -670,6 +674,46 @@ impl EnvExt for Ext {
         Ok(self.context.message_context.current().id())
     }
 
+    fn pay_rent(&mut self, program_id: ProgramId, rent: u128) -> Result<u128, Self::Error> {
+        let block_count = u32::try_from(rent / self.context.rent_cost).unwrap_or(u32::MAX);
+
+        let old_paid_blocks = self
+            .context
+            .program_rents
+            .get(&program_id)
+            .copied()
+            .unwrap_or(0);
+
+        let (paid_blocks, blocks_to_pay) = match old_paid_blocks.overflowing_add(block_count) {
```

Comment:
I'd also return error on overflow of u32
for case of handling such thing on pallet in overflow case I'd refund programs balance